src/main/java/edu/wpi/grip/core/Connection.java

```diff
@@ -29,6 +30,10 @@ public Connection(EventBus eventBus, Socket<? extends T> outputSocket, Socket<T>
         checkNotNull(inputSocket);
         checkNotNull(outputSocket);
         checkNotNull(eventBus);
+        checkArgument(!Socket.Direction.INPUT.equals(outputSocket.getDirection()),
+                "outputSocket cannot be an input socket");
+        checkArgument(!Socket.Direction.OUTPUT.equals(inputSocket.getDirection()),
+                "inputSocket cannot be an output socket");
```

Comment:
That would allow you to change this:
```java
    @Subscribe
    public void onOutputChanged(SocketChangedEvent e) {
        if (e.getSocket() == outputSocket) {
            inputSocket.setValue(outputSocket.getValue());
        }
    }
```

To be something more like this:
```java
    @Subscribe
    public void onOutputChanged(SocketChangedEvent e) {
        if (e.getSocket() == outputSocket) {
            inputSocket.connect(outputSocket);
        }
    }
```
Allowing you to make the `setValue` function package protected on SocketInput.